File: SQLVec-pg/batched query/SQLVec.py
import psycopg2
import time
import numpy as np
import os
import struct
import argparse

N_DIM = 128
DATA_SIZE = 4

def fvecs_read(filename, c_contiguous=True, record_count=-1, line_offset=0, record_dtype=np.int32):
    if record_count > 0:
        record_count *= N_DIM + 1
    if line_offset > 0:
        line_offset *= (N_DIM + 1) * DATA_SIZE
    fv = np.fromfile(filename, dtype=record_dtype, count=record_count, offset=line_offset)
    if fv.size == 0:
        return np.zeros((0, 0))
    dim = fv.view(np.int32)[0]
    assert dim > 0
    fv = fv.reshape(-1, 1 + dim)
    if not all(fv.view(np.int32)[:, 0] == dim):
        raise IOError("Non-uniform vector sizes in " + filename)
    fv = fv[:, 1:]
    if c_contiguous:
        fv = fv.copy()
    return fv

def load_truthset(bin_file,k):
    actual_file_size = os.path.getsize(bin_file)

    with open(bin_file, 'rb') as f:
        npts_i32 = struct.unpack('i', f.read(4))[0]
        dim_i32 = struct.unpack('i', f.read(4))[0]
        npts = npts_i32
        dim = dim_i32

        truthset_type = -1  # 1 means truthset has ids and distances, 2 means only ids, -1 is error
        expected_file_size_with_dists = 2 * npts * dim * 4 + 2 * 4
        expected_file_size_just_ids = npts * dim * 4 + 2 * 4

        if actual_file_size == expected_file_size_with_dists:
            truthset_type = 1
        elif actual_file_size == expected_file_size_just_ids:
            truthset_type = 2

        if truthset_type == -1:
            raise ValueError(f"Error. File size mismatch. File should have bin format, with "
                             f"npts followed by ngt followed by npts*ngt ids and optionally "
                             f"followed by npts*ngt distance values; actual size: "
                             f"{actual_file_size}, expected: {expected_file_size_with_dists} or "
                             f"{expected_file_size_just_ids}")

        ids = np.fromfile(f, dtype=np.uint32, count=npts * dim)
        dists = None

        if truthset_type == 1:
            dists = np.fromfile(f, dtype=np.float32, count=npts * dim)

    return ids.reshape(npts, k)

def main(args):
    conn = psycopg2.connect(database=args.db, user="postgres", password="229229", host="127.0.0.1", port="5432")
    conn.autocommit = True
    cur = conn.cursor()
    cur.execute('CREATE EXTENSION IF NOT EXISTS vector')
    
    start_time_total = time.time()
    # init node
    cur.execute('SELECT data FROM data_table WHERE id = %s', (args.enter_node,))
    enter_node = cur.fetchone()
    # L_table,V_table
    cur.execute('CREATE TEMP TABLE L_table (id INTEGER, dist DOUBLE PRECISION, batch INTEGER)')
    cur.execute('INSERT INTO L_table SELECT %s, %s <-> data as dist, num AS batch FROM query_table', (args.enter_node, enter_node))
    cur.execute('CREATE TEMP TABLE V_table (id INTEGER, batch INTEGER)')
    cur.execute('CREATE INDEX vid ON V_table USING btree (id,batch);')
    
    # the core search process
    cnt = 9
    it = 0
    while cnt != 0:
        it += 1
        print(it)
        cur.execute('CREATE TEMP TABLE tp1 AS SELECT L_table.* FROM L_table LEFT JOIN V_table ON L_table.id = V_table.id AND L_table.batch = V_table.batch WHERE V_table.id IS NULL;')
        cur.execute('SELECT count(*) FROM tp1')
        cnt = cur.fetchone()[0]
  
        cur.execute('CREATE TEMP TABLE tpv AS SELECT t1.* FROM tp1 t1 JOIN (SELECT batch, min(dist) AS mind FROM tp1 GROUP BY batch) t2 ON t1.batch = t2.batch AND t1.dist = t2.mind')

        cur.execute('CREATE TEMP TABLE tp2 AS SELECT batch, UNNEST(nbs) as neighbor FROM tpv JOIN Graph_index_table ON tpv.id = Graph_index_table.id')
        
        cur.execute('INSERT INTO L_table SELECT neighbor AS id, data_table.data <-> query_table.data AS dist, tp2.batch FROM data_table, query_table, tp2 WHERE tp2.neighbor = data_table.id AND query_table.num = tp2.batch')

        cur.execute('INSERT INTO V_table SELECT id,batch FROM tpv')
        cur.execute('DROP TABLE tp1')
        cur.execute('DROP TABLE tp2')
        cur.execute('DROP TABLE tpv')

        cur.execute('CREATE TEMP TABLE tp_table AS SELECT d.id, d.dist, d.batch FROM (SELECT l1.*, row_number() OVER (PARTITION BY batch ORDER BY dist) AS class_rank FROM (SELECT DISTINCT * FROM L_table) l1) d WHERE d.class_rank <= %s ORDER BY batch, dist',(args.l,))
        cur.execute('DROP TABLE L_table')
        cur.execute('ALTER TABLE tp_table RENAME TO L_table')
    end_time_total = time.time()

    # 查询结果
    cur.execute('SELECT d.id, d.batch FROM (SELECT l1.*, row_number() OVER (PARTITION BY batch ORDER BY dist) AS class_rank FROM L_table l1) d WHERE d.class_rank <= %s ORDER BY batch, dist',(args.k,))
    res_all = cur.fetchall()

    conn.commit()
    conn.close()
    
    # to load groundtruth vectors
    if args.gt_type == '.fvecs':
        ground_truth = fvecs_read(args.gt_path)
    elif args.gt_type == '.bin':
        ground_truth = load_truthset(args.gt_path, args.k)
    
    ground_truth = ground_truth[:,:args.k]

    # to get the result vector of each batch
    batch_dict = {}
    for i in range(len(res_all)):
        if res_all[i][1] in batch_dict:
            batch_dict[res_all[i][1]].append(res_all[i][0])
        else:
            batch_dict[res_all[i][1]] = [res_all[i][0]]
        
    result = list(batch_dict.values())

    intersection = 0
    for i in range(len(result)):
        intersection += len(np.intersect1d(ground_truth[i], result[i]))
    
    print("Iteration: " + str(it))
    print("Average Recall: " + str(intersection/(args.k*args.querysize)))
    print("Total Time Taken: " + str(end_time_total - start_time_total))
    print("QPS: " + str(1/(end_time_total - start_time_total)*args.querysize))
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='batched query')
    parser.add_argument('--db', type=str, default='sift')
    parser.add_argument('--enter_node', type=int, default=123742)
    parser.add_argument('--k', type=int, default=100, help='top-k search, k number')
    parser.add_argument('--l', type=int, default=100, help='the search list length')
    parser.add_argument('--querysize', type=int, default=10000, help='the query size of datasets') 
    parser.add_argument('--gt_path', type=str, help='the path of groundtruth file')
    parser.add_argument('--gt_type', type=str, default='.fvecs', help='the type of groundtruth file')
    args = parser.parse_args()

    main(args)

File: SQLVec-db/pq_intergation_cpp/SQLVec.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <cassert>
#include <duckdb.hpp>
#include <algorithm>
#include <unordered_map>
#include <chrono>

#define N_DIM 128
#define DATA_SIZE 4

std::vector<float> fvecs_read(const std::string &filename, bool c_contiguous = true, int record_count = -1, int line_offset = 0, const std::string &record_dtype = "int32") {
    std::ifstream input(filename, std::ios::binary);
    if (!input) {
        std::cerr << "Error opening file: " << filename << std::endl;
        exit(1);
    }

    if (record_count > 0) {
        record_count *= N_DIM + 1;
    }
    if (line_offset > 0) {
        line_offset *= (N_DIM + 1) * DATA_SIZE;
    }

    input.seekg(line_offset, std::ios::beg);
    std::vector<float> fv(record_count);
    input.read(reinterpret_cast<char*>(fv.data()), record_count * sizeof(float));
    input.close();

    if (fv.size() == 0) {
        return std::vector<float>(0);
    }

    int dim = static_cast<int>(fv[0]);
    assert(dim > 0);
    fv.resize(fv.size() / (dim + 1) * dim);

    if (c_contiguous) {
        std::vector<float> contiguous_fv(fv.begin(), fv.end());
        return contiguous_fv;
    }
    return fv;
}

int main(int argc, char** argv) {
    std::string database = "/home/anns/ANNS/duckdb/sift.duckdb";  
    int enter_node = 123742;
    std::string groundtruth_path = "/home/anns/ANNS/dataset/sift1m/sift_groundtruth.ivecs";
    int L = 100;
    int K = 100;

    // Connect to DuckDB database
    duckdb::DuckDB db(database);
    duckdb::Connection con(db);

    // Create table
    // index table
    con.Query("CREATE TABLE temp_table AS SELECT * FROM read_csv('/home/zzq229/ANNSonSQL/tables/graph_index_vamana_R70_L75_A12.csv')");
    con.Query("CREATE TABLE Graph_index_table AS select id, cast(neighbor as int[]) as neighbor from temp_table");
    con.Query("DROP TABLE temp_table"); 
    // data_table
    con.Query("CREATE TABLE temp_table AS SELECT * FROM read_csv('/home/zzq229/ANNSonSQL/tables/data_table.csv')");
    con.Query("CREATE TABLE data_table AS select id, cast(data as int[]) as data from temp_table");
    con.Query("DROP TABLE temp_table");
    // query_table
    con.Query("CREATE TABLE temp_table AS SELECT * FROM read_csv('/home/zzq229/ANNSonSQL/tables/query_table.csv')");
    con.Query("CREATE TABLE query_table AS select num, cast(data as int[]) as data from temp_table");
    con.Query("DROP TABLE temp_table");
    // Read ground truth data
    // std::vector<float> ground_truth = fvecs_read(groundtruth_path);

    // Create indices
    con.Query("CREATE INDEX g_id_index ON Graph_index_table(id)");
    con.Query("CREATE INDEX d_id_index ON data_table(id)");

    // Enable optimizer
    con.Query("PRAGMA enable_optimizer");

    // auto enter_node = con.Query("SELECT data FROM data_table WHERE id = 123742")->FetchRaw();
    // std::cout << enter_node << std::endl;

    // Start query processing
    auto start_time_total = std::chrono::high_resolution_clock::now();

    con.Query("CREATE TEMP TABLE L_table(id INTEGER, dist DOUBLE, batch INTEGER)");

    // std::unique_ptr<duckdb::PreparedStatement> p1 = con.Prepare("INSERT INTO L_table SELECT 123742, list_distance([20, 8, 19, 10, 13, 27, 23, 9, 58, 45, 34, 18, 17, 27, 33, 58, 27, 51, 85, 28, 18, 38, 34, 67, 18, 7, 32, 38, 63, 38, 7, 3, 52, 21, 27, 22, 39, 29, 37, 22, 86, 51, 62, 22, 57, 60, 54, 62, 52, 34, 37, 21, 60, 64, 102, 101, 60, 56, 66, 20, 36, 55, 33, 40, 44, 24, 40, 20, 40, 35, 35, 23, 93, 50, 62, 31, 54, 53, 81, 44, 80, 67, 100, 22, 78, 56, 58, 50, 43, 14, 42, 17, 40, 53, 59, 54, 12, 5, 28, 40, 63, 12, 13, 13, 42, 21, 21, 24, 36, 39, 47, 33, 66, 23, 21, 18, 25, 31, 51, 73, 25, 14, 25, 20, 33, 26, 18, 15], data) as dist, num as batch from query_table");
    // p1->Execute();
    con.Query("INSERT INTO L_table SELECT 123742, list_distance([20,8,19,10,13,27,23,9,58,45,34,18,17,27,33,58,27,51,85,28,18,38,34,67,18,7,32,38,63,38,7,3,52,21,27,22,39,29,37,22,86,51,62,22,57,60,54,62,52,34,37,21,60,64,102,101,60,56,66,20,36,55,33,40,44,24,40,20,40,35,35,23,93,50,62,31,54,53,81,44,80,67,100,22,78,56,58,50,43,14,42,17,40,53,59,54,12,5,28,40,63,12,13,13,42,21,21,24,36,39,47,33,66,23,21,18,25,31,51,73,25,14,25,20,33,26,18,15], data) as dist, num as batch from query_table");

    con.Query("CREATE TEMP TABLE V_table(id INTEGER, dist DOUBLE, batch INTEGER)");

    int cnt = 9;
    int it = 0;
    while (cnt != 0) {
        it++;
        // Step 1
        con.Query("CREATE TEMP TABLE tp1 AS SELECT * FROM L_table EXCEPT SELECT * FROM V_table");
        cnt = con.Query("SELECT * FROM tp1")->RowCount();
        // std::cout << cnt << std::endl;

        // Step 1+2
        con.Query("CREATE TEMP TABLE tpv AS SELECT DISTINCT t1.* FROM tp1 t1 JOIN (SELECT batch, min(dist) AS mind FROM tp1 GROUP BY batch) t2 ON t1.batch = t2.batch AND t1.dist = t2.mind");

        // Step 3
        con.Query("CREATE TEMP TABLE tp2 AS SELECT DISTINCT batch, unnest(neighbor) as neighbor FROM tpv JOIN Graph_index_table USING(id)");

        con.Query("INSERT INTO L_table SELECT neighbor as id, list_distance(data_table.data, query_table.data) as dist, tp2.batch FROM data_table, query_table, tp2 WHERE tp2.neighbor = data_table.id AND query_table.num = tp2.batch");

        con.Query("INSERT INTO V_table SELECT * FROM tpv");
        con.Query("DROP TABLE tp1");
        con.Query("DROP TABLE tp2");
        con.Query("DROP TABLE tpv");

        con.Query("CREATE TABLE tp_table AS SELECT DISTINCT d.id, d.dist, d.batch FROM (SELECT l1.*, row_number() over (PARTITION BY batch ORDER BY dist) AS class_rank FROM (SELECT DISTINCT * FROM L_table) l1) d WHERE d.class_rank <= 100 ORDER BY batch, dist");
        con.Query("DROP TABLE L_table");
        con.Query("ALTER TABLE tp_table RENAME TO L_table");
    }

    auto end_time_total = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed_time = end_time_total - start_time_total;

    // Collect results
    // auto res_all = con.Query("SELECT d.id, d.batch FROM (SELECT l1.*, row_number() over (PARTITION BY batch ORDER BY dist) AS class_rank FROM L_table l1) d WHERE d.class_rank <= 100 ORDER BY batch, dist")
    // std::unordered_map<int, std::vector<int>> batch_dict;

    // for (size_t i = 0; i < res_all.size(); i++) {
    //     int id = res_all[i][0];
    //     int batch = res_all[i][1];
    //     batch_dict[batch].push_back(id);
    // }

    // std::vector<std::vector<int>> result;
    // for (const auto &entry : batch_dict) {
    //     result.push_back(entry.second);
    // }

    // // Calculate average recall
    // int intersection = 0;
    // for (size_t i = 0; i < result.size(); i++) {
    //     std::vector<int> &res = result[i];
    //     std::vector<int> &truth = ground_truth[i];
    //     std::vector<int> inter;
    //     std::set_intersection(res.begin(), res.end(), truth.begin(), truth.end(), std::back_inserter(inter));
    //     intersection += inter.size();
    // }

    std::cout << "Iteration: " << it << std::endl;
    // std::cout << "Average Recall: " << static_cast<double>(intersection) / (100 * 10000) << std::endl;
    std::cout << "Total Time Taken: " << elapsed_time.count() << " seconds" << std::endl;
    std::cout << "Average Time Taken: " << elapsed_time.count() / 10000 << " seconds" << std::endl;

    return 0;
}
